core/src/main/java/org/kohsuke/stapler/export/Property.java

```diff
@@ -199,118 +198,126 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
 
         Class c = value.getClass();
 
-        Model model;
-        try {
-            model = owner.get(c, parent.type, name);
-        } catch (NotExportableException ex) {
-            if(STRING_TYPES.contains(c)) {
-                writer.value(value.toString());
-                return;
-            }
-            if(PRIMITIVE_TYPES.contains(c)) {
-                writer.valuePrimitive(value);
-                return;
+        if (c.getAnnotation(ExportedBean.class) == null) {
+            handleNotExportable(expected, value, pruner, writer, skipIfFail, c);
+        } else {
+            try {
+                writer.type(expected, value.getClass());
+            } catch (AbstractMethodError _) {
+                // legacy impl that doesn't understand it
             }
-            Class act = c.getComponentType();
-            if (act !=null) { // array
-                Range r = pruner.getRange();
-                writer.startArray();
-                if (value instanceof Object[]) {
-                    // typical case
-                    for (Object item : r.apply((Object[]) value)) {
-                        writeBuffered(act, item, pruner, writer);
-                    }
-                } else {
-                    // more generic case
-                    int len = Math.min(r.max, Array.getLength(value));
-                    for (int i=r.min; i<len; i++) {
-                        writeBuffered(act, Array.get(value, i), pruner, writer);
-                    }
+            try {
+                Model model = owner.getOrNull(c, parent.type, name);
+                if (model == null) {
+                    throw new NotExportableException(c);
                 }
-                writer.endArray();
+                writer.startObject();
+                model.writeNestedObjectTo(value, pruner, writer);
+                writer.endObject();
+            } catch (NotExportableException ex) {
+                handleNotExportable(expected, value, pruner, writer, skipIfFail, c);
```

Comment:
[reduced diff](https://github.com/stapler/stapler/pull/127/files?w=1) is easier to read here (generally dislike patches that change indentation of large existing code blocks but sometimes it is hard to avoid)